### PR TITLE
BREAKING CHANGE: SASjs CLI 2.0 - Electric Boogaloo

### PR DESCRIPTION
**Pull request by krishna-acondy**
BREAKING CHANGE: SASjs CLI 2.0 Release

## Issue

Fixes https://github.com/sasjs/cli/issues/269
Fixes https://github.com/sasjs/cli/issues/185
Fixes https://github.com/sasjs/cli/issues/165
Fixes https://github.com/sasjs/cli/issues/287
Fixes https://github.com/sasjs/cli/issues/302
Fixes https://github.com/sasjs/cli/issues/358

## Intent

* Rewrite the CLI to improve stability, reduce bugs and enhance developer experience.
* Make the implementation work with the new `Target` model defined in `@sasjs/utils`.
* Improve the stability of the unit tests.
* Separate the server-based tests from mocked ones.

### Targets
* There is now ONE single way of retrieving a target - `findTargetInConfiguration` in `src/utils/config`.
* It returns an object containing a `target` which is of type `Target` from `@sasjs/utils`, and a boolean `isLocal` which indicates whether the target is from the local config or the global RC file.
* `Target`s are read-only. That means the only way to modify them is by changing the values in the JSON files, or by replacing existing targets with new ones.
* When a target is saved to a config file, we call the `target.toJson()` method on it, which will convert it into a JSON representation suitable for storage.
* When retrieving a target, the JSON is used to create a `Target` object.

### Project Architecture
* Each command has its own folder within `src/commands`.
* Inside that folder, there will be one module that will export the main function(s) for that command.
* There is a subfolder within that folder called `internal`, which will contain any helper functions specific to this command.
* We should strive to split functionality out into multiple modules so we don't end up with 1000+ line files.
* Splitting out functions also gives us the ability to mock out entire modules using Jest when testing our code.

### TypeScript
* All code and tests have now been migrated over to TypeScript
* New interfaces and types have been added where applicable.

### Error Handling
* The individual command implementations will throw any errors that they encounter.
* The error will be handled in `main.ts`.
* The functions in `main.ts` will return a `ReturnCode` to the CLI, which will be used as the exit code.
* The different `ReturnCode`s available are `Success`, `InvalidCommand` and `InternalError`.
* We will be able to use this mechanism to show friendly messages to make things clearer to the user.
* There is still some iffy code in some of the commands which we need to refactor.
* All command implementations should be throwing errors as soon as possible in the process - failing fast is important and will help guard against getting into inconsistent states.

### Logging
* A `Logger` (from `@sasjs/utils`) is now instantiated at the very beginning when any command is run.
* The logger is available on the `process` object as `process.logger`.
* The log level is based on a `LOG_LEVEL` env variable and defaults to `Error` (might make sense to default to `Debug` or `Info` in the CLI).
* Most instances of using `console.log` along with `chalk` have been refactored to use the new logger.
* `sasjs help` is still using `chalk` since it is actually required there to make the text more easily readable.

### Testing
* Most of the functions from `helpers.js` which were being attached to the `global` object have been removed.
* Some have been moved into the `utils/test` module. This module and all associated files are excluded from the built package.
* The separate `test` folder has been removed, and the tests for each command are in a `spec` folder alongside the actual command.
* All the tests have been refactored to create their test projects within the current folder rather than at the root level.
* This will reduce the likelihood of clashes between different test suites.
* All test suites also use unique timestamp-based names for their targets and apps.
* The `spec` files are also excluded from the final bundle.
* There are utilities added to `utils/test` for creating and tearing down test apps and targets.
* There are now two separate `jest` configurations - one for mocked/non-server tests, and one for server-based tests.
* When creating server-based tests, you will need to create files called `<name>.spec.server.ts`.
* When creating non-server tests, you will need to create files called `<name>.spec.ts`.
* The CI workflows have been updated to run the 'smoke tests' first.
* Then, they'll run all the non-server tests - this takes less than a minute.
* If all of them pass, they'll run the server-based tests, which may take upto 10 minutes.

### Compile / Build / Deploy
* I've separated these into their own individual commands to split out the code better and make them independent.
* The compilation check is now done in the `compile` module - so we won't recompile if not required.
* The `build` command will call compile, which will recompile if necessary.
* When running `cb` or `cbd` commands, there is now error handling at each step of the process - so if something fails on the `compile` step, we won't attempt to execute any further steps and so on.

phew

## Checks

- [x] Code is formatted correctly (`npm run lint:fix`).
- [x] Any new functionality has been unit tested.
- [x] All unit tests are passing (`npm test`).
- [x] All CI checks are green.
- [x] JSDoc comments have been added or updated.
- [x] Reviewer is assigned.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -42,7 +44,7 @@
   "dependencies": {
     "@sasjs/adapter": "^1.19.4",
     "@sasjs/core": "^1.12.3",
```
Comment (allanbowe) [inline package.json:46]: This can go to v2 now

File: src/cli.spec.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      appName,
+      `/Public/app/cli-tests/${appName}`
+    )
+    done()
+  })
+
+  afterEach(async (done) => {
+    await removeTestApp(__dirname, target.name)
+    await removeFromGlobalConfig(target.name)
+    jest.clearAllMocks()
```
Comment (YuryShkoda) [inline src/cli.spec.ts:30]: Do we still need to clear it here (we've set `clearMocks: true` in jest config)?

File: src/commands/compile/internal/config.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      configuration.jobConfig &&
+      configuration.jobConfig.initProgram
+    ) {
+      jobInitContent = await readFile(
+        path.join(buildSourceFolder, configuration.jobConfig.initProgram)
+      )
+    }
+  }
+
+  console.log('JOBINIT: ', jobInitContent)
```
Comment (YuryShkoda) [inline src/commands/compile/internal/config.ts:83]: We don't need this `console.log()`

File: src/utils/config.ts [hunk 4]
```diff
@@ -1,26 +1,29 @@
...
-    Promise.resolve(null)
-  })
+/**
+ * Returns an object that represents the SASjs CLI configuration in a given file.
+ * @param {string} pathToFile - the path to the file in question.
+ * @returns {Configuration} configuration object if available.
+ */
+export async function getConfiguration(
+  pathToFile: string
+): Promise<Configuration> {
```
Comment (YuryShkoda) [inline src/utils/config.ts:18]: this function doesn't return a Promise

File: src/utils/config.ts [hunk 5]
```diff
@@ -30,70 +33,73 @@ export async function getConfiguration(pathToFile) {
...
  * @param {boolean} viyaSpecific - will fall back to the first target of type SASVIYA.
+ * @returns {Target} target or fallback when one is found.
  */
 export async function findTargetInConfiguration(
-  targetName,
+  targetName: string,
   viyaSpecific = false
-) {
-  const projectRoot = await getProjectRoot()
+): Promise<{ target: Target; isLocal: boolean }> {
```
Comment (YuryShkoda) [inline src/utils/config.ts:41]: this function doesn't return a Promise

**Review by YuryShkoda — DISMISSED** [review]

File: package.json (hunk 1 above)
Comment (krishna-acondy) [reply package.json]: Done!

File: src/commands/compile/internal/config.ts (hunk 3 above)
Comment (krishna-acondy) [reply src/commands/compile/internal/config.ts]: Removed.

File: src/utils/config.ts (hunk 5 above)
Comment (krishna-acondy) [reply src/utils/config.ts:41]: `async/await` in ES6/TypeScript is based on `Promise`s, so this function _does_ actually return a `Promise`.

File: src/utils/config.ts (hunk 4 above)
Comment (krishna-acondy) [reply src/utils/config.ts:18]: `async/await` in ES6/TypeScript is based on `Promise`s, so this function _does_ actually return a `Promise`.

**Review by medjedovicm — APPROVED** [review]

**Review by saadjutt01 — DISMISSED** [review]

File: src/utils/config.ts [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+    path.join(projectRoot, 'sasjs', 'sasjsconfig.json')
+  ).catch(() => null)
+  if (localConfig && localConfig.programFolders) {
+    programFolders = programFolders.concat(localConfig.programFolders)
+  }
+
+  const { target } = await findTargetInConfiguration(targetName)
+
+  if (!target) {
+    throw new Error(
```
Comment (saadjutt01) [inline src/utils/config.ts:263]: this piece of code is unreachable (and redundant as code throws in `findTargetInConfiguration` if there's no target.

File: src/utils/config.ts [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+    path.join(projectRoot, 'sasjs', 'sasjsconfig.json')
+  ).catch(() => null)
+  if (localConfig && localConfig.programFolders) {
+    macroFolders = macroFolders.concat(localConfig.programFolders)
+  }
+
+  const { target } = await findTargetInConfiguration(targetName)
+
+  if (!target) {
+    throw new Error(
```
Comment (saadjutt01) [inline src/utils/config.ts:300]: this piece of code is unreachable (and redundant as code throws in `findTargetInConfiguration` if there's no target.

**Review by saadjutt01 — CHANGES_REQUESTED** [review]

File: src/utils/config.ts (hunk 6 above)
Comment (krishna-acondy) [reply src/utils/config.ts]: I've removed that bit of code.

File: src/utils/config.ts (hunk 7 above)
Comment (krishna-acondy) [reply src/utils/config.ts]: I've removed that bit of code.

File: src/cli.spec.ts (hunk 2 above)
Comment (krishna-acondy) [reply src/cli.spec.ts]: Keeping it there just in case.

**Review by saadjutt01 — APPROVED** [review]

**Pull request merged**